crates/core_simd/src/swizzle.rs

```diff
@@ -364,4 +364,131 @@ where
 
         (Even::swizzle2(self, other), Odd::swizzle2(self, other))
     }
+
+    /// Takes a slice of a vector to produce a shorter vector.
+    ///
+    /// This is equivalent to computing `&self[OFFSET..OFFSET+LEN]` on
+    /// the underlying array.
+    ///
+    /// ```
+    /// # #![feature(portable_simd)]
+    /// # use core::simd::Simd;
+    /// let x = Simd::from_array([0, 1, 2, 3, 4, 5, 6, 7]);
+    /// let y = x.slice::<2, 4>();
+    /// assert_eq!(y.to_array(), [2, 3, 4, 5]);
+    /// ```
+    ///
+    /// Will be rejected at compile time if `OFFSET + LEN > LANES`.
+    #[inline]
+    #[must_use = "method returns a new vector and does not mutate the original inputs"]
+    pub fn slice<const OFFSET: usize, const LEN: usize>(self) -> Simd<T, LEN>
+    where
+        LaneCount<LEN>: SupportedLaneCount,
+    {
+        const fn slice_index<const LEN: usize>(offset: usize, lanes: usize) -> [usize; LEN] {
+            assert!(offset + LEN <= lanes, "slice out of bounds");
+            let mut index = [0; LEN];
+            let mut i = 0;
+            while i < LEN {
+                index[i] = i + offset;
+                i += 1;
+            }
+            index
+        }
+        struct Slice<const OFFSET: usize>;
+        impl<const OFFSET: usize, const LEN: usize, const LANES: usize> Swizzle<LANES, LEN>
+            for Slice<OFFSET>
+        {
+            const INDEX: [usize; LEN] = slice_index::<LEN>(OFFSET, LANES);
+        }
+        Slice::<OFFSET>::swizzle(self)
+    }
+
+    /// Concatenates two vectors of equal length.
+    ///
+    /// Due to limitations in const generics, the length of the resulting vector cannot be inferred
+    /// from the input vectors. You must specify it explicitly.
+    ///
+    /// ```
+    /// # #![feature(portable_simd)]
+    /// # use core::simd::Simd;
+    /// let x = Simd::from_array([0, 1, 2, 3]);
+    /// let y = Simd::from_array([4, 5, 6, 7]);
+    /// let z = x.concat_to::<8>(y);
+    /// assert_eq!(z.to_array(), [0, 1, 2, 3, 4, 5, 6, 7]);
+    /// ```
+    ///
+    /// Will be rejected at compile time if `LANES * 2 != DOUBLE_LANES`.
+    #[inline]
+    #[must_use = "method returns a new vector and does not mutate the original inputs"]
+    pub fn concat_to<const DOUBLE_LANES: usize>(self, other: Self) -> Simd<T, DOUBLE_LANES>
+    where
+        LaneCount<DOUBLE_LANES>: SupportedLaneCount,
+    {
+        const fn concat_index<const DOUBLE_LANES: usize>(lanes: usize) -> [Which; DOUBLE_LANES] {
+            assert!(lanes * 2 == DOUBLE_LANES);
```

Comment:
Just checking: Are you aware that when you put something in a `const fn`, and then the `const fn` is evaluated at runtime, that this means an `assert!` in it will not be evaluated at compilation time, but rather will be evaluated at runtime?

Reply:
it's only used at compile time, so that doesn't matter here.

Reply:
Ah, right, I see now, it happens in the associated constant. Then that requires monomorphization which puts it at risk of depending on trait and const evaluation details. Hmm. I do want to allow us to use certain post-monomorphization errors but I currently don't fully understand the evaluation patterns that rustc will do to intuit in what cases this will/will not happen.